# Number conversion - Tumor Mapping - Basem Provided Logic 

In [1]:
# Importing necessary libraries for data processing, analysis, or visualization
import pandas as pd
# p_idx=pd.read_excel(r'C:\Users\basem\OneDrive\Desktop\shakaib\Tumor module\TTL pkg\TTL_updated_Master_File_Dec_2024\Reference_files\Master_files\TTL_post_processing_12_6_24.xlsx',header=None)
# p_num=pd.read_excel(r'C:\Users\basem\OneDrive\Desktop\shakaib\Tumor module\TTL pkg\TTL_updated_Master_File_Dec_2024\Reference_files\Master_files\TTL_post_processing_raw_11_28_23.xlsx',header=None)
# master=pd.read_excel(r'C:\Users\basem\OneDrive\Desktop\shakaib\Tumor module\TTL pkg\TTL_updated_Master_File_Dec_2024\Reference_files\Master_files\master_file_12_6_24.xlsx')

p_idx=pd.read_excel(r'./TTL_post_processing_12_6_24.xlsx',header=None)
p_num=pd.read_excel(r'.\TTL_post_processing_raw_11_28_23.xlsx',header=None)
master=pd.read_excel(r'.\master_file_12_6_24.xlsx')

In [2]:
master.head(60)

L1                                                 L2  \
0   Leukemia                                                NaN   
1        NaN                         Acute Myelogenous Leukemia   
2        NaN                                                NaN   
3        NaN                                                NaN   
4        NaN                                                NaN   
5        NaN                                                NaN   
6        NaN                                                NaN   
7        NaN                                                NaN   
8        NaN                                                NaN   
9        NaN                                                NaN   
10       NaN                                                NaN   
11       NaN                                                NaN   
12       NaN                       Acute Lymphoblastic Leukemia   
13       NaN                                                NaN   
14       NaN                                                NaN   
15       NaN                                                NaN   
16       NaN                                                NaN   
17       NaN                                                NaN   
18       NaN                       Chronic Lymphocytic Leukemia   
19       NaN                       Chronic Myelogenous Leukemia   
20       NaN                                Hairy Cell Leukemia   
21       NaN                               Plasma Cell Leukemia   
22       NaN                                 Mast Cell Leukemia   
23       NaN                        Acute Eosinophilic leukemia   
24       NaN                          Acute Basophilic Leukemia   
25       NaN  Acute Undifferentiated or Biphenotypic (Mixed)...   
26       NaN  T-cell Large Granular Lymphocytic Leukemia (TLGL)   
27       NaN                            Prolymphocytic Leukemia   
28       NaN                                                NaN   
29       NaN                                                NaN   
30       NaN                                 stem cell leukemia   
31       NaN  blastic plasmacytoid dendritic cell neoplasm (...   
32       NaN                                   nk-cell leukemia   
33       NaN                                     leukemia cutis   
34       NaN                               hypoplastic leukemia   
35       NaN                    monoclonal b-cell lymphocytosis   
36       NaN                                                NaN   
37  Lymphoma                                                NaN   
38       NaN                                   Hodgkin Lymphoma   
39       NaN                                                NaN   
40       NaN                                                NaN   
41       NaN                                                NaN   
42       NaN                                                NaN   
43       NaN                                                NaN   
44       NaN                               Non-Hodgkin Lymphoma   
45       NaN                                                NaN   
46       NaN                                                NaN   
47       NaN                                                NaN   
48       NaN                                                NaN   
49       NaN                                                NaN   
50       NaN                                                NaN   
51       NaN                                                NaN   
52       NaN                                                NaN   
53       NaN                                                NaN   
54       NaN                                                NaN   
55       NaN                                                NaN   
56       NaN                                                NaN   
57       NaN                                                NaN   
58       NaN                                             

In [4]:
# Outer loop to repeat the process 4 times, possibly for ensuring complete data alignment.
for j in range(4):
    print(j)  # Print the iteration number for debugging or tracking progress.
    
    # Iterate through each row of the 'master' DataFrame, focusing on columns L1 to L5.
    for idx, row in master[["L1", "L2", "L3", "L4", "L5"]].iterrows():
        
        # Process column 'L4': Check if the value in L4 is not a NaN (float NaN is used for missing data).
        if not isinstance(row.L4, float):
            # Identify the first index where 'L5' is null after the current row (idx).
            l3i = master["L5"].iloc[idx + 1:].isnull()
            l3i = l3i[l3i].index.values[0]  # Get the index of the first null value in L5.
            
            # Update all rows between the current index (idx) and the identified null index (l3i).
            for i in range(idx, l3i):
                master.at[i, "L4"] = master.iloc[idx, 3]  # Assign the value of L4 from the current row.

        # Process column 'L3': Similar logic applied to propagate values in column 'L3'.
        if not isinstance(row.L3, float):
            l3i = master["L4"].iloc[idx + 1:].isnull()
            l3i = l3i[l3i].index.values[0]
            for i in range(idx, l3i):
                master.at[i, "L3"] = master.iloc[idx, 2]

        # Process column 'L2': Similar logic applied to propagate values in column 'L2'.
        if not isinstance(row.L2, float):
            l3i = master["L3"].iloc[idx + 1:].isnull()
            l3i = l3i[l3i].index.values[0]
            for i in range(idx, l3i):
                master.at[i, "L2"] = master.iloc[idx, 1]

        # # Process column 'L1': Similar logic applied to propagate values in column 'L1'.
        # if not isinstance(row.L1, float):
        #     l3i = master["L2"].iloc[idx + 1:].isnull()
        #     l3i = l3i[l3i].index.values[0]
        #     for i in range(idx, l3i):
        #         master.at[i, "L1"] = master.iloc[idx, 0]

# Display the updated 'master' DataFrame.
master


0
1
2
3


L1                          L2  \
0                      Leukemia                         NaN   
1                      Leukemia  Acute Myelogenous Leukemia   
2                           NaN  Acute Myelogenous Leukemia   
3                           NaN  Acute Myelogenous Leukemia   
4                           NaN  Acute Myelogenous Leukemia   
...                         ...                         ...   
1172   malignant pilonidal cyst                         NaN   
1173                        NaN                         NaN   
1174   paraneoplastic syndromes                         NaN   
1175                        NaN                         NaN   
1176  subcutaneous malignancies                         NaN   

                                         L3   L4   L5     stage_range  \
0                                       NaN  NaN  NaN  ['No staging']   
1                                       NaN  NaN  NaN  ['No staging']   
2     AML with minimal differentiation (M0)  NaN  NaN  ['No staging']   
3               AML without maturation (M1)  NaN  NaN  ['No staging']   
4                  AML with maturation (M2)  NaN  NaN  ['No staging']   
...                                     ...  ...  ...             ...   
1172                                    NaN  NaN  NaN    [1, 2, 3, 4]   
1173                                    NaN  NaN  NaN             NaN   
1174                                    NaN  NaN  NaN  ['no staging']   
1175                                    NaN  NaN  NaN             NaN   
1176                                    NaN  NaN  NaN    [1, 2, 3, 4]   

                                 tumor_type Unnamed: 7  
0                                       NaN        NaN  
1                acute myelogenous leukemia        NaN  
2     aml with minimal differentiation (m0)        NaN  
3               aml without maturation (m1)        NaN  
4                  aml with maturation (m2)        NaN  
...                                     ...        ...  
1172                                    NaN        NaN  
1173                                    NaN        NaN  
1174                                    NaN        NaN  
1175                                    NaN        NaN  
1176                                    NaN        NaN  

[1177 rows x 8 columns]

In [ ]:
master.head(60)

L1                                                 L2  \
0   Leukemia                                                NaN   
1   Leukemia                         Acute Myelogenous Leukemia   
2        NaN                         Acute Myelogenous Leukemia   
3        NaN                         Acute Myelogenous Leukemia   
4        NaN                         Acute Myelogenous Leukemia   
5        NaN                         Acute Myelogenous Leukemia   
6        NaN                         Acute Myelogenous Leukemia   
7        NaN                         Acute Myelogenous Leukemia   
8        NaN                         Acute Myelogenous Leukemia   
9        NaN                         Acute Myelogenous Leukemia   
10       NaN                         Acute Myelogenous Leukemia   
11       NaN                         Acute Myelogenous Leukemia   
12       NaN                       Acute Lymphoblastic Leukemia   
13       NaN                       Acute Lymphoblastic Leukemia   
14       NaN                       Acute Lymphoblastic Leukemia   
15       NaN                       Acute Lymphoblastic Leukemia   
16       NaN                       Acute Lymphoblastic Leukemia   
17       NaN                       Acute Lymphoblastic Leukemia   
18       NaN                       Chronic Lymphocytic Leukemia   
19       NaN                       Chronic Myelogenous Leukemia   
20       NaN                                Hairy Cell Leukemia   
21       NaN                               Plasma Cell Leukemia   
22       NaN                                 Mast Cell Leukemia   
23       NaN                        Acute Eosinophilic leukemia   
24       NaN                          Acute Basophilic Leukemia   
25       NaN  Acute Undifferentiated or Biphenotypic (Mixed)...   
26       NaN  T-cell Large Granular Lymphocytic Leukemia (TLGL)   
27       NaN                            Prolymphocytic Leukemia   
28       NaN                            Prolymphocytic Leukemia   
29       NaN                            Prolymphocytic Leukemia   
30       NaN                                 stem cell leukemia   
31       NaN  blastic plasmacytoid dendritic cell neoplasm (...   
32       NaN                                   nk-cell leukemia   
33       NaN                                     leukemia cutis   
34       NaN                               hypoplastic leukemia   
35       NaN                    monoclonal b-cell lymphocytosis   
36       NaN                                                NaN   
37  Lymphoma                                                NaN   
38  Lymphoma                                   Hodgkin Lymphoma   
39       NaN                                   Hodgkin Lymphoma   
40       NaN                                   Hodgkin Lymphoma   
41       NaN                                   Hodgkin Lymphoma   
42       NaN                                   Hodgkin Lymphoma   
43       NaN                                   Hodgkin Lymphoma   
44       NaN                               Non-Hodgkin Lymphoma   
45       NaN                               Non-Hodgkin Lymphoma   
46       NaN                               Non-Hodgkin Lymphoma   
47       NaN                               Non-Hodgkin Lymphoma   
48       NaN                               Non-Hodgkin Lymphoma   
49       NaN                               Non-Hodgkin Lymphoma   
50       NaN                               Non-Hodgkin Lymphoma   
51       NaN                               Non-Hodgkin Lymphoma   
52       NaN                               Non-Hodgkin Lymphoma   
53       NaN                               Non-Hodgkin Lymphoma   
54       NaN                               Non-Hodgkin Lymphoma   
55       NaN                               Non-Hodgkin Lymphoma   
56       NaN                               Non-Hodgkin Lymphoma   
57       NaN                               Non-Hodgkin Lymphoma   
58       NaN                               Non-Hodgkin Ly

In [6]:
# Remove leading and trailing whitespace from string values in column 'c' of the DataFrame,
# leaving non-string values unchanged.
for c in ["L1", "L2", "L3", "L4", "L5"]:
    master[c] = master[c].apply(lambda x: x.strip() if isinstance(x, str) else x)

In [7]:
# General operation or computation
master.head(60)

L1                                                 L2  \
0   Leukemia                                                NaN   
1   Leukemia                         Acute Myelogenous Leukemia   
2        NaN                         Acute Myelogenous Leukemia   
3        NaN                         Acute Myelogenous Leukemia   
4        NaN                         Acute Myelogenous Leukemia   
5        NaN                         Acute Myelogenous Leukemia   
6        NaN                         Acute Myelogenous Leukemia   
7        NaN                         Acute Myelogenous Leukemia   
8        NaN                         Acute Myelogenous Leukemia   
9        NaN                         Acute Myelogenous Leukemia   
10       NaN                         Acute Myelogenous Leukemia   
11       NaN                         Acute Myelogenous Leukemia   
12       NaN                       Acute Lymphoblastic Leukemia   
13       NaN                       Acute Lymphoblastic Leukemia   
14       NaN                       Acute Lymphoblastic Leukemia   
15       NaN                       Acute Lymphoblastic Leukemia   
16       NaN                       Acute Lymphoblastic Leukemia   
17       NaN                       Acute Lymphoblastic Leukemia   
18       NaN                       Chronic Lymphocytic Leukemia   
19       NaN                       Chronic Myelogenous Leukemia   
20       NaN                                Hairy Cell Leukemia   
21       NaN                               Plasma Cell Leukemia   
22       NaN                                 Mast Cell Leukemia   
23       NaN                        Acute Eosinophilic leukemia   
24       NaN                          Acute Basophilic Leukemia   
25       NaN  Acute Undifferentiated or Biphenotypic (Mixed)...   
26       NaN  T-cell Large Granular Lymphocytic Leukemia (TLGL)   
27       NaN                            Prolymphocytic Leukemia   
28       NaN                            Prolymphocytic Leukemia   
29       NaN                            Prolymphocytic Leukemia   
30       NaN                                 stem cell leukemia   
31       NaN  blastic plasmacytoid dendritic cell neoplasm (...   
32       NaN                                   nk-cell leukemia   
33       NaN                                     leukemia cutis   
34       NaN                               hypoplastic leukemia   
35       NaN                    monoclonal b-cell lymphocytosis   
36       NaN                                                NaN   
37  Lymphoma                                                NaN   
38  Lymphoma                                   Hodgkin Lymphoma   
39       NaN                                   Hodgkin Lymphoma   
40       NaN                                   Hodgkin Lymphoma   
41       NaN                                   Hodgkin Lymphoma   
42       NaN                                   Hodgkin Lymphoma   
43       NaN                                   Hodgkin Lymphoma   
44       NaN                               Non-Hodgkin Lymphoma   
45       NaN                               Non-Hodgkin Lymphoma   
46       NaN                               Non-Hodgkin Lymphoma   
47       NaN                               Non-Hodgkin Lymphoma   
48       NaN                               Non-Hodgkin Lymphoma   
49       NaN                               Non-Hodgkin Lymphoma   
50       NaN                               Non-Hodgkin Lymphoma   
51       NaN                               Non-Hodgkin Lymphoma   
52       NaN                               Non-Hodgkin Lymphoma   
53       NaN                               Non-Hodgkin Lymphoma   
54       NaN                               Non-Hodgkin Lymphoma   
55       NaN                               Non-Hodgkin Lymphoma   
56       NaN                               Non-Hodgkin Lymphoma   
57       NaN                               Non-Hodgkin Lymphoma   
58       NaN                               Non-Hodgkin Ly

In [8]:
# Identify rows in the DataFrame where all the columns 'L1', 'L2', 'L3', 'L4', 'L5' are null (NaN).
# The condition checks if the sum of nulls across these columns is equal to 5 (i.e., all are null).
parent_b = master[master[["L1", "L2", "L3", "L4", "L5"]].isnull().apply(sum, axis=1) == 5].index.values

# Loop through the identified indices (parent_b) to process the DataFrame.
for idx, e in enumerate(parent_b):
    if idx == 0:
        s = 0  # For the first element, start from the beginning of the DataFrame.
    else:
        s = parent_b[idx-1] + 1  # For subsequent elements, start after the previous index.
    
    # Print the current index and the start (s) and end (e) positions.
    print(idx, f's={s}, e={e}')
    
    # Retrieve the value from 'L1' at the start index (s).
    word = master.L1.values[s]
    
    # Update 'L1' column values for all rows between start (s) and end (e) with the retrieved word.
    for j in range(s, e):
        master.at[j, "L1"] = word


0 s=0, e=36
1 s=37, e=143
2 s=144, e=152
3 s=153, e=174
4 s=175, e=200
5 s=201, e=247
6 s=248, e=273
7 s=274, e=294
8 s=295, e=304
9 s=305, e=311
10 s=312, e=313
11 s=314, e=325
12 s=326, e=342
13 s=343, e=350
14 s=351, e=366
15 s=367, e=371
16 s=372, e=385
17 s=386, e=402
18 s=403, e=430
19 s=431, e=434
20 s=435, e=445
21 s=446, e=449
22 s=450, e=458
23 s=459, e=466
24 s=467, e=473
25 s=474, e=477
26 s=478, e=515
27 s=516, e=580
28 s=581, e=614
29 s=615, e=707
30 s=708, e=724
31 s=725, e=745
32 s=746, e=761
33 s=762, e=835
34 s=836, e=847
35 s=848, e=860
36 s=861, e=868
37 s=869, e=874
38 s=875, e=893
39 s=894, e=922
40 s=923, e=945
41 s=946, e=967
42 s=968, e=1020
43 s=1021, e=1030
44 s=1031, e=1036
45 s=1037, e=1047
46 s=1048, e=1072
47 s=1073, e=1085
48 s=1086, e=1090
49 s=1091, e=1094
50 s=1095, e=1099
51 s=1100, e=1109
52 s=1110, e=1115
53 s=1116, e=1118
54 s=1119, e=1130
55 s=1131, e=1132
56 s=1133, e=1134
57 s=1135, e=1147
58 s=1148, e=1154
59 s=1155, e=1156
60 s=1157, e=1158
6

In [ ]:
display(p_idx.head())
display(p_num.iloc[600:].head(60))

0                           1  \
0  Leukemia                         NaN   
1       NaN  Acute Myelogenous Leukemia   
2       NaN                         NaN   
3       NaN                         NaN   
4       NaN                         NaN   

                                       2    3    4  
0                                    NaN  NaN  NaN  
1                                    NaN  NaN  NaN  
2  AML with minimal differentiation (M0)  NaN  NaN  
3            AML without maturation (M1)  NaN  NaN  
4               AML with maturation (M2)  NaN  NaN

0    1   2  \
600                      hematological or solid tumors  BL1 NaN   
601                         malignant tumors lymphomas  BL1 NaN   
602                     colorectal neoplasms malignant  BL1 NaN   
603                   non-ovarian gynecological cancer  BL1 NaN   
604                     malignant hematologic diseases  BL1 NaN   
605                             hematopoietic neoplasm  BL1 NaN   
606                            salivary duct carcinoma  BL1 NaN   
607                 primary or secondary myelofibrosis  BL1 NaN   
608                                         lungcancer  BL1 NaN   
609                                             tnbc -  BL1 NaN   
610                       malignant intracranial tumor  BL1 NaN   
611                                composite lymphomas  BL1 NaN   
612                                               rare  BL1 NaN   
613                        lymphoproliferative disease  BL1 NaN   
614                                        endometrial  BL1 NaN   
615                                 mast cell disorder  BL1 NaN   
616                               pancreatic neoplasia  BL1 NaN   
617         carcinomas of the esophagogastric junction  BL1 NaN   
618                             periampullary neoplasm  BL1 NaN   
619                             pseudomyxoma peritonei  BL1 NaN   
620                              malignant solid tumor  BL1 NaN   
621                              biliary system tumors  BL1 NaN   
622                               cancer cervix uterus  BL1 NaN   
623                 malignant epithelioid mesothelioma  BL1 NaN   
624                  cancer of kidney and renal pelvis  BL1 NaN   
625       myeloproliferative/myelodysplastic neoplasms  BL1 NaN   
626                            musculoskeletal cancers  BL1 NaN   
627                               cancer head and neck  BL1 NaN   
628  neoplasm of uncertain malignant potential of t...  BL1 NaN   
629                                     neuroendocrine  BL1 NaN   
630               chronic myeloproliferative neoplasms  BL1 NaN   
631                        epithelial tumor, malignant  BL1 NaN   
632                                    carcinoma;blood  BL1 NaN   
633                               broncho-lung cancers  BL1 NaN   
634                                  neoplasm, hepatic  BL1 NaN   
635                                     primary tumour  BL1 NaN   
636                                             colon-  BL1 NaN   
637                         cutaneous malignant cancer  BL1 NaN   
638                                               mrcc  BL1 NaN   
639  biliary tract (intrahepatic, extrahepatic chol...  BL1 NaN   
640                           lateral spreading tumors  BL1 NaN   
641                       neoplastic colorectal tissue  BL1 NaN   
642                                     neck neoplasms  BL1 NaN   
643                               melanoma of the skin  BL1 NaN   
644                           uc (ureter and urethera)  BL1 NaN   
645                            latent cancer (prostate  BL1 NaN   
646                              salivary gland tumors  BL1 NaN   
647                                            current  BL1 NaN   
648                                 cutaneous neoplasm  BL1 NaN   
649                            colon and rectum cancer  BL1 NaN   
650               hematologic and lymphocytic disorder  BL1 NaN   
651               malignant neoplasms of urinary tract  BL1 NaN   
652                      germ cell and embryonal tumor  BL1 NaN   
653                              genitourinary cancers  BL1 NaN   
654                     carcinoma of the head and neck  BL1 NaN   
655                               complex liver tumors  BL1 NaN   
656                                   'brain neoplasms  BL1 NaN   
657        squamous cell carcinoma of the head or neck  BL1 NaN   
658                   gastro-intestinal stromal tumors  BL1 NaN   
659  malignant lesion of the dist

In [10]:
# Add a new column 'idx' to the DataFrame 'p_idx' with values ranging from 1 to the length of 'p_idx'.
p_idx['idx'] = list(range(1, len(p_idx) + 1))

# Reorder the columns of 'p_idx' so that 'idx' becomes the first column, followed by columns 0, 1, 2, 3, and 4.
p_idx = p_idx[['idx', 0, 1, 2, 3, 4]]

# Display the modified 'p_idx' DataFrame.
p_idx

idx                                                  0  \
0        1                                           Leukemia   
1        2                                                NaN   
2        3                                                NaN   
3        4                                                NaN   
4        5                                                NaN   
...    ...                                                ...   
1130  1131                                        igg myeloma   
1131  1132                                        igd myeloma   
1132  1133                                      Fibroxanthoma   
1133  1134  uterine adenocarcinoma (includes serous, clear...   
1134  1135  monoclonal gammopathy (includes MGUS, MGNS, MGRS)   

                               1                                      2    3  \
0                            NaN                                    NaN  NaN   
1     Acute Myelogenous Leukemia                                    NaN  NaN   
2                            NaN  AML with minimal differentiation (M0)  NaN   
3                            NaN            AML without maturation (M1)  NaN   
4                            NaN               AML with maturation (M2)  NaN   
...                          ...                                    ...  ...   
1130                         NaN                                    NaN  NaN   
1131                         NaN                                    NaN  NaN   
1132                         NaN                                    NaN  NaN   
1133                         NaN                                    NaN  NaN   
1134                         NaN                                    NaN  NaN   

        4  
0     NaN  
1     NaN  
2     NaN  
3     NaN  
4     NaN  
...   ...  
1130  NaN  
1131  NaN  
1132  NaN  
1133  NaN  
1134  NaN  

[1135 rows x 6 columns]

In [11]:
# General operation - value at index 60
p_idx.head(60)

idx         0                                                  1  \
0     1  Leukemia                                                NaN   
1     2       NaN                         Acute Myelogenous Leukemia   
2     3       NaN                                                NaN   
3     4       NaN                                                NaN   
4     5       NaN                                                NaN   
5     6       NaN                                                NaN   
6     7       NaN                                                NaN   
7     8       NaN                                                NaN   
8     9       NaN                                                NaN   
9    10       NaN                                                NaN   
10   11       NaN                                                NaN   
11   12       NaN                                                NaN   
12   13       NaN                       Acute Lymphoblastic Leukemia   
13   14       NaN                                                NaN   
14   15       NaN                                                NaN   
15   16       NaN                                                NaN   
16   17       NaN                                                NaN   
17   18       NaN                                                NaN   
18   19       NaN                       Chronic Lymphocytic Leukemia   
19   20       NaN                       Chronic Myelogenous Leukemia   
20   21       NaN                                Hairy Cell Leukemia   
21   22       NaN                               Plasma Cell Leukemia   
22   23       NaN                                 Mast Cell Leukemia   
23   24       NaN                        Acute Eosinophilic leukemia   
24   25       NaN                          Acute Basophilic Leukemia   
25   26       NaN  Acute Undifferentiated or Biphenotypic (Mixed)...   
26   27       NaN  T-cell Large Granular Lymphocytic Leukemia (TLGL)   
27   28       NaN                            Prolymphocytic Leukemia   
28   29       NaN                                                NaN   
29   30       NaN                                                NaN   
30   31       NaN                                 stem cell leukemia   
31   32       NaN  blastic plasmacytoid dendritic cell neoplasm (...   
32   33  Lymphoma                                                NaN   
33   34       NaN                                   Hodgkin Lymphoma   
34   35       NaN                                                NaN   
35   36       NaN                                                NaN   
36   37       NaN                                                NaN   
37   38       NaN                                                NaN   
38   39       NaN                                                NaN   
39   40       NaN                               Non-Hodgkin Lymphoma   
40   41       NaN                                                NaN   
41   42       NaN                                                NaN   
42   43       NaN                                                NaN   
43   44       NaN                                                NaN   
44   45       NaN                                                NaN   
45   46       NaN                                                NaN   
46   47       NaN                                                NaN   
47   48       NaN                                                NaN   
48   49       NaN                                                NaN   
49   50       NaN                                                NaN   
50   51       NaN                                                NaN   
51   52       NaN                                                NaN   
52   53       NaN                                                NaN   
53   54       NaN                                                NaN   
54   55       NaN                           

In [12]:
def fn_assignment_numbers(x):
    try:
        # Convert the input 'x' to a string, remove specific patterns ('!=', '!-'), and strip spaces.
        x = str(x).replace('!=', '').replace("!-", '').replace(' ', '').strip()
        
        # Split the string on commas to create a list of values.
        x = x.split(',')
        
        # Convert the list of string numbers into integers.
        x = [int(i) for i in x]
        
        # Filter rows in 'p_idx' where the 'idx' column matches the numbers in 'x'.
        # Extract the values of columns 0, 1, 2, 3, 4, and flatten them into a 1D array.
        x1 = p_idx[p_idx['idx'].isin(x)][[0, 1, 2, 3, 4]].values.flatten()
        
        # If there are valid values in 'x1', clean them:
        if len(x1) > 0:
            # Convert elements to strings, strip spaces, and exclude 'nan' values.
            x1 = [str(i).strip() for i in x1 if str(i) != 'nan']
        
        # Return the cleaned list of values.
        return x1
    except:
        # If an error occurs, return None.
        return None

# Apply the 'fn_assignment_numbers' function to column 3 of the 'p_num' DataFrame
# and store the results in a new column called 'nums'.
p_num['nums'] = p_num[3].apply(fn_assignment_numbers)

# Display the modified 'p_num' DataFrame.
p_num

0    1   2                     3  \
0                    spinal cord tumor ( s  BL1 NaN                   426   
1     malignant neoplasm in the peritoneum  BL1 NaN              608, 274   
2                            biliary tract  BL1 NaN                   594   
3                hematologic malignancies/  BL1 NaN  1, 33, 320, 330, 625   
4           gastroenteropancreatico tumors  BL1 NaN                   806   
...                                    ...  ...  ..                   ...   
8417            Carcinoma of Urinary Tract  NaN NaN                   NaN   
8418                  urothelial carcinoma  NaN NaN                   NaN   
8419                      urothelial tumor  NaN NaN                   NaN   
8420                     urothelial cancer  NaN NaN                   NaN   
8421              Uterine Corpus Carcinoma  NaN NaN                   NaN   

        4    5    6    7    8  \
0     NaN  NaN  NaN  NaN  NaN   
1     NaN  NaN  NaN  NaN  NaN   
2     NaN  NaN  NaN  NaN  NaN   
3     NaN  NaN  NaN  NaN  NaN   
4     NaN  NaN  NaN  NaN  NaN   
...   ...  ...  ...  ...  ...   
8417  NaN  NaN  NaN  NaN  NaN   
8418  NaN  NaN  NaN  NaN  NaN   
8419  NaN  NaN  NaN  NaN  NaN   
8420  NaN  NaN  NaN  NaN  NaN   
8421  NaN  NaN  NaN  NaN  NaN   

                                                   nums  
0     [Central Nervous System Cancer (brain and spin...  
1     [Primary Peritoneal Cancer (for psuedomxyoma p...  
2     [Biliary Tract Cancer (includes cholangiocarci...  
3     [Leukemia, Lymphoma, Plasma Cell Dyscrasia (in...  
4        [gastroenteropancreatic neuroendocrine tumors]  
...                                                 ...  
8417                                               None  
8418                                               None  
8419                                               None  
8420                                               None  
8421                                               None  

[8422 rows x 10 columns]

In [13]:
# Select only the columns 'L1', 'L2', 'L3', 'L4', and 'L5' from the 'master' DataFrame.
# Reset the index of the DataFrame, creating a new default integer index and adding the old index as a column.
master = master[["L1", "L2", "L3", "L4", "L5"]].reset_index()

# Increment the 'index' column (created during the reset) by 1.
master['index'] = master['index'].apply(lambda x: x + 1)

# Display the modified 'master' DataFrame.
master

index                         L1                          L2  \
0         1                   Leukemia                         NaN   
1         2                   Leukemia  Acute Myelogenous Leukemia   
2         3                   Leukemia  Acute Myelogenous Leukemia   
3         4                   Leukemia  Acute Myelogenous Leukemia   
4         5                   Leukemia  Acute Myelogenous Leukemia   
...     ...                        ...                         ...   
1172   1173   malignant pilonidal cyst                         NaN   
1173   1174                        NaN                         NaN   
1174   1175   paraneoplastic syndromes                         NaN   
1175   1176                        NaN                         NaN   
1176   1177  subcutaneous malignancies                         NaN   

                                         L3   L4   L5  
0                                       NaN  NaN  NaN  
1                                       NaN  NaN  NaN  
2     AML with minimal differentiation (M0)  NaN  NaN  
3               AML without maturation (M1)  NaN  NaN  
4                  AML with maturation (M2)  NaN  NaN  
...                                     ...  ...  ...  
1172                                    NaN  NaN  NaN  
1173                                    NaN  NaN  NaN  
1174                                    NaN  NaN  NaN  
1175                                    NaN  NaN  NaN  
1176                                    NaN  NaN  NaN  

[1177 rows x 6 columns]

In [14]:
# Filter the DataFrame to include only rows where 'L1' is equal to 'Myeloproliferative Neoplasm'.
master[master.L1=="Myeloproliferative Neoplasm"].head(60)

index                           L1  \
894    895  Myeloproliferative Neoplasm   
895    896  Myeloproliferative Neoplasm   
896    897  Myeloproliferative Neoplasm   
897    898  Myeloproliferative Neoplasm   
898    899  Myeloproliferative Neoplasm   
899    900  Myeloproliferative Neoplasm   
900    901  Myeloproliferative Neoplasm   
901    902  Myeloproliferative Neoplasm   
902    903  Myeloproliferative Neoplasm   
903    904  Myeloproliferative Neoplasm   
904    905  Myeloproliferative Neoplasm   
905    906  Myeloproliferative Neoplasm   
906    907  Myeloproliferative Neoplasm   
907    908  Myeloproliferative Neoplasm   
908    909  Myeloproliferative Neoplasm   
909    910  Myeloproliferative Neoplasm   
910    911  Myeloproliferative Neoplasm   
911    912  Myeloproliferative Neoplasm   
912    913  Myeloproliferative Neoplasm   
913    914  Myeloproliferative Neoplasm   
914    915  Myeloproliferative Neoplasm   
915    916  Myeloproliferative Neoplasm   
916    917  Myeloproliferative Neoplasm   
917    918  Myeloproliferative Neoplasm   
918    919  Myeloproliferative Neoplasm   
919    920  Myeloproliferative Neoplasm   
920    921  Myeloproliferative Neoplasm   
921    922  Myeloproliferative Neoplasm   

                                                    L2  \
894                                                NaN   
895                          Essential Thrombocythemia   
896                                  Polycythemia Vera   
897                                      Myelofibrosis   
898                                      Myelofibrosis   
899                                      Myelofibrosis   
900                                      Myelofibrosis   
901                                      Myelofibrosis   
902                  atypical chronic myeloid leukemia   
903                           Chronic Myeloid Leukemia   
904  myeloid/lymphoid neoplasms with eosinophilia a...   
905  myeloid/lymphoid neoplasms with eosinophilia a...   
906  myeloid/lymphoid neoplasms with eosinophilia a...   
907  myeloid/lymphoid neoplasms with eosinophilia a...   
908  myeloid/lymphoid neoplasms with eosinophilia a...   
909  myeloid/lymphoid neoplasms with eosinophilia a...   
910  myeloid/lymphoid neoplasms with eosinophilia a...   
911  myeloid/lymphoid neoplasms with eosinophilia a...   
912  myeloid/lymphoid neoplasms with eosinophilia a...   
913  myeloid/lymphoid neoplasms with eosinophilia a...   
914  myeloid/lymphoid neoplasms with eosinophilia a...   
915                    Chronic Myelomonocytic leukemia   
916                    Chronic Myelomonocytic leukemia   
917                    Chronic Myelomonocytic leukemia   
918                    Chronic Myelomonocytic leukemia   
919                   Juvenile Myelomonocytic leukemia   
920                         chronic monocytic leukemia   
921                      chronic neutrophilic leukemia   

                                             L3  \
894                                         NaN   
895                                         NaN   
896                                         NaN   
897                                         NaN   
898                       primary myelofibrosis   
899                     secondary myelofibrosis   
900                     secondary myelofibrosis   
901                     secondary myelofibrosis   
902                                         NaN   
903                                         NaN   
904                                         NaN   
905                                Mastocytosis   
906                                Mastocytosis   
907                                Mastocytosis   
908                                Mastocytosis   
909                                Mastocytosis   
910                                Mastocytosis   
911                                Mastocytosis   
912                                Mastocytosis   
913         chronic eosinophilic leukemia (nos)   
914

In [15]:
# Create a reference to the 'master' DataFrame named 'master2'. 
master2=master

In [16]:
# Select and display only the columns '0' and 'nums' from the 'p_num' DataFrame.
p_num[[0, 'nums']]

# Retrieve and display the 97th row (index 96) of the 'p_num' DataFrame for inspection.
p_num.iloc[96]

0       endocervical cancer
1                       BL1
2                       NaN
3                       491
4                       NaN
5                       NaN
6                       NaN
7                       NaN
8                       NaN
nums      [Cervical Cancer]
Name: 96, dtype: object

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>index</th>
      <th>L1</th>
      <th>L2</th>
      <th>L3</th>
      <th>L4</th>
      <th>L5</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>879</th>
      <td>880</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>880</th>
      <td>881</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>Essential Thrombocythemia</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>881</th>
      <td>882</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>Polycythemia Vera</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>882</th>
      <td>883</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>Myelofibrosis</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>883</th>
      <td>884</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>Myelofibrosis</td>
      <td>primary myelofibrosis</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>884</th>
      <td>885</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>Myelofibrosis</td>
      <td>secondary myelofibrosis</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>885</th>
      <td>886</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>Myelofibrosis</td>
      <td>secondary myelofibrosis</td>
      <td>post-essential thrombocythemia myelofibrosis</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>886</th>
      <td>887</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>Myelofibrosis</td>
      <td>secondary myelofibrosis</td>
      <td>post-polycythemia vera myelofibrosis</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>887</th>
      <td>888</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>atypical chronic myeloid leukemia</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>888</th>
      <td>889</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>Chronic Myeloid Leukemia</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>889</th>
      <td>890</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>myeloid/lymphoid neoplasms with eosinophilia a...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>890</th>
      <td>891</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>myeloid/lymphoid neoplasms with eosinophilia a...</td>
      <td>Mastocytosis</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>891</th>
      <td>892</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>myeloid/lymphoid neoplasms with eosinophilia a...</td>
      <td>Mastocytosis</td>
      <td>Indolent Systemic Mastocytosis</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>892</th>
      <td>893</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>myeloid/lymphoid neoplasms with eosinophilia a...</td>
      <td>Mastocytosis</td>
      <td>Smoldering systemic mastocytosis</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>893</th>
      <td>894</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>myeloid/lymphoid neoplasms with eosinophilia a...</td>
      <td>Mastocytosis</td>
      <td>Cutaneous Mastocytosis</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>894</th>
      <td>895</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>myeloid/lymphoid neoplasms with eosinophilia a...</td>
      <td>Mastocytosis</td>
      <td>Aggressive Systemic Mastocytosis</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>895</th>
      <td>896</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>myeloid/lymphoid neoplasms with eosinophilia a...</td>
      <td>Mastocytosis</td>
      <td>mast cell sarcoma</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>896</th>
      <td>897</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>myeloid/lymphoid neoplasms with eosinophilia a...</td>
      <td>Mastocytosis</td>
      <td>Systemic Mastocytosis with Associated Hematolo...</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>897</th>
      <td>898</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>myeloid/lymphoid neoplasms with eosinophilia a...</td>
      <td>Mastocytosis</td>
      <td>Systemic Mastocytosis with Associated Hematolo...</td>
      <td>mastocytosis associated with myeloproliferativ...</td>
    </tr>
    <tr>
      <th>898</th>
      <td>899</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>myeloid/lymphoid neoplasms with eosinophilia a...</td>
      <td>chronic eosinophilic leukemia (nos)</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>899</th>
      <td>900</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>myeloid/lymphoid neoplasms with eosinophilia a...</td>
      <td>Primary Hypereosinophilic Syndrome</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>900</th>
      <td>901</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>Chronic Myelomonocytic leukemia</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>901</th>
      <td>902</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>Chronic Myelomonocytic leukemia</td>
      <td>Chronic myelomonocytic leukemia-1 (CMML-1)</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>902</th>
      <td>903</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>Chronic Myelomonocytic leukemia</td>
      <td>Chronic myelomonocytic leukemia-2 (CMML-2)</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>903</th>
      <td>904</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>Chronic Myelomonocytic leukemia</td>
      <td>Chronic myelomonocytic leukemia-2 (CMML-2)</td>
      <td></td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>904</th>
      <td>905</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>Juvenile Myelomonocytic leukemia</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>905</th>
      <td>906</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>chronic monocytic leukemia</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>906</th>
      <td>907</td>
      <td>Myeloproliferative Neoplasm</td>
      <td>chronic neutrophilic leukemia</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>

In [17]:
# Extract the value from the 'nums' column of the 97th row (index 96) in the 'p_num' DataFrame.
p_num.iloc[96]['nums']

['Cervical Cancer']

In [18]:
# Initialize a new column 'Alternate_words' in the 'master2' DataFrame and set all its values to None.
master2['Alternate_words'] = None

# Iterate through each row of the 'master2' DataFrame.
for idx, row in master2.iterrows():
    try:
        # Drop NaN values from the current row and convert it to a dictionary for easy key-value access.
        row = row.dropna().to_dict()
        
        # Initialize an empty dictionary to store alternate words for each column.
        d = {}
        
        # Retrieve values of L1, L2, L3, L4, and L5 from the row.
        l1, l2, l3, l4, l5 = row.get('L1'), row.get('L2'), row.get('L3'), row.get('L4'), row.get('L5')
        
        # Check if L1 exists and find its alternate words.
        if l1:
            # Create a copy of relevant columns from 'p_num', removing rows with NaN values.
            alt_df = p_num[[0, 'nums']].copy().dropna()
            # Filter rows in 'alt_df' where the 'nums' column contains the L1 value.
            alt_df = alt_df[alt_df['nums'].apply(lambda x: True if l1 in x else False)]
            # Store the L1 value and its alternate words in the dictionary.
            d["L1"] = [l1] + alt_df[0].values.tolist()
        
        # Check if L2 exists and find its alternate words.
        if l2:
            alt_df = p_num[[0, 'nums']].copy().dropna()
            alt_df = alt_df[alt_df['nums'].apply(lambda x: True if l2 in x else False)]
            d["L2"] = [l2] + alt_df[0].unique().tolist()
        
        # Check if L3 exists and find its alternate words.
        if l3:
            alt_df = p_num[[0, 'nums']].copy().dropna()
            alt_df = alt_df[alt_df['nums'].apply(lambda x: True if l3 in x else False)]
            d["L3"] = [l3] + alt_df[0].values.tolist()
        
        # Check if L4 exists and find its alternate words.
        if l4:
            alt_df = p_num[[0, 'nums']].copy().dropna()
            alt_df = alt_df[alt_df['nums'].apply(lambda x: True if l4 in x else False)]
            d["L4"] = [l4] + alt_df[0].values.tolist()
        
        # Check if L5 exists and find its alternate words.
        if l5:
            alt_df = p_num[[0, 'nums']].copy().dropna()
            alt_df = alt_df[alt_df['nums'].apply(lambda x: True if l5 in x else False)]
            d["L5"] = [l5] + alt_df[0].values.tolist()

        # Assign the dictionary of alternate words to the 'Alternate_words' column for the current row.
        master2.at[idx, "Alternate_words"] = d
    
    # Handle any exceptions that occur during the iteration and print the index where it happened.
    except Exception as e:
        print(e, idx)

# Display the modified 'master2' DataFrame.
master2

index                         L1                          L2  \
0         1                   Leukemia                         NaN   
1         2                   Leukemia  Acute Myelogenous Leukemia   
2         3                   Leukemia  Acute Myelogenous Leukemia   
3         4                   Leukemia  Acute Myelogenous Leukemia   
4         5                   Leukemia  Acute Myelogenous Leukemia   
...     ...                        ...                         ...   
1172   1173   malignant pilonidal cyst                         NaN   
1173   1174                        NaN                         NaN   
1174   1175   paraneoplastic syndromes                         NaN   
1175   1176                        NaN                         NaN   
1176   1177  subcutaneous malignancies                         NaN   

                                         L3   L4   L5  \
0                                       NaN  NaN  NaN   
1                                       NaN  NaN  NaN   
2     AML with minimal differentiation (M0)  NaN  NaN   
3               AML without maturation (M1)  NaN  NaN   
4                  AML with maturation (M2)  NaN  NaN   
...                                     ...  ...  ...   
1172                                    NaN  NaN  NaN   
1173                                    NaN  NaN  NaN   
1174                                    NaN  NaN  NaN   
1175                                    NaN  NaN  NaN   
1176                                    NaN  NaN  NaN   

                                        Alternate_words  
0     {'L1': ['Leukemia', 'hematologic malignancies/...  
1     {'L1': ['Leukemia', 'hematologic malignancies/...  
2     {'L1': ['Leukemia', 'hematologic malignancies/...  
3     {'L1': ['Leukemia', 'hematologic malignancies/...  
4     {'L1': ['Leukemia', 'hematologic malignancies/...  
...                                                 ...  
1172  {'L1': ['malignant pilonidal cyst', 'any nonmy...  
1173                                                 {}  
1174  {'L1': ['paraneoplastic syndromes', 'paraneopl...  
1175                                                 {}  
1176  {'L1': ['subcutaneous malignancies', 'sub-cuta...  

[1177 rows x 7 columns]

In [21]:
# General operation or computation
master2[master2.L1=='Central Nervous System Cancer (brain and spine tumors)']['Alternate_words'].values[1]

{'L1': ['Central Nervous System Cancer (brain and spine tumors)',
  'spinal cord tumor ( s',
  'cancer of any type',
  'cerebellar neoplasm',
  'solid-organ cancer',
  'malignant or nonmalignant disease',
  "pregnancy related' 'cancer",
  'extremities neoplasms',
  'malignant solid neoplasm',
  'any solid malignancy',
  'all  solid tumors',
  'non hematologic cancer',
  'childhood malignancy',
  'brainstem neoplasm',
  'haemato-oncologic',
  'intracranial neoplasm',
  'rare solid tumors',
  'malignant neoplasm in the brain',
  'any nonmyeloid tumor',
  'childhood brain tumors',
  'hematology and oncology',
  "'other solid tumors",
  'solid and hematological malignancies',
  'malignant disorders',
  'nonematologic malignancies',
  'spinal neoplasm',
  'other solid cancers',
  'brain neoplasms, malignant',
  'malignant brain tumor',
  'all cancer',
  'brain',
  'tumors of any histology',
  'malignant neoplasm in the central nervous system',
  'nonematological solid tumors',
  'neoplasms 

In [23]:
import re  # Import the 're' module for regular expression operations.

# Convert the values in column 3 of the 'p_num' DataFrame to strings.
p_num[3] = p_num[3].apply(str)

# Filter the 'p_num' DataFrame to include only rows where column 3 contains the character '!' (indicating a "not equal" pattern).
p_num_with_not_equal = p_num[p_num[3].str.contains('!')]

# Display the filtered 'p_num_with_not_equal' DataFrame.
p_num_with_not_equal

# Define a function to extract and process "not equal" numbers from a string.
def fn_get_not_equal_num(x):
    # Convert the input 'x' to a string.
    x = str(x)
    
    # Use a regular expression to find all numbers following a pattern like "! =", "!=-", "! -", etc.
    # The pattern matches "! [=-] number" (e.g., !=5, !- 12).
    x = re.findall(r'! ?[=-] ?(\d+)', x)
    
    # Convert the extracted numbers to integers.
    x = [int(i) for i in x]
    
    # Use the extracted numbers to filter rows in 'p_idx' where the 'idx' column matches the numbers.
    # Extract and flatten the values from columns 0, 1, 2, 3, and 4 in the filtered rows.
    x1 = p_idx[p_idx['idx'].isin(x)][[0, 1, 2, 3, 4]].values.flatten()
    
    # Clean the resulting list by removing 'nan' values or empty strings.
    if len(x1) > 0:
        x1 = [i for i in x1 if str(i) != 'nan' or str(i).strip() != ""]
    
    # Return the cleaned list of alternate words.
    return x1

# Apply the 'fn_get_not_equal_num' function to the '3' column of the 'p_num_with_not_equal' DataFrame.
# Create a new column 'not_equal' to store the extracted and processed alternate words.
p_num_with_not_equal['not_equal'] = p_num_with_not_equal[3].apply(fn_get_not_equal_num)

# Display the updated 'p_num_with_not_equal' DataFrame.
p_num_with_not_equal


C:\Users\user\AppData\Local\Temp\ipykernel_16688\4005760301.py:37: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  p_num_with_not_equal['not_equal'] = p_num_with_not_equal[3].apply(fn_get_not_equal_num)


0    1   2  \
95                                   brainstem neoplasm  BL1 NaN   
107                               intracranial neoplasm  BL1 NaN   
118                     malignant neoplasm in the brain  BL1 NaN   
125                              childhood brain tumors  BL1 NaN   
185                          brain neoplasms, malignant  BL1 NaN   
...                                                 ...  ...  ..   
7560  non-germinal center b-cell-like subtypes of di...  BL4 NaN   
7570                                      non-gcb dlbcl  BL4 NaN   
7793            non aids-related,non-hodgkin's lymphoma  BL4 NaN   
7937                non follicular non hodgkin lymphoma  BL4 NaN   
8041                 non-gastric marginal zone lymphoma  BL4 NaN   

                                                      3     4    5    6    7  \
95    426, 63, !=430, !=434, !=446, !=448, !=453, !=...   NaN  NaN  NaN  NaN   
107          63, 426, !=471, !=485, !=807, !=430, !=470  anat  NaN  NaN  NaN   
118          426, 63, !=430, !=465, !=470, !=471, !=485   NaN  NaN  NaN  NaN   
125          426, 63, !=430, !=465, !=470, !=471, !=485   NaN  NaN  NaN  NaN   
185          426, 63, !=430, !=465, !=470, !=471, !=485   NaN  NaN  NaN  NaN   
...                                                 ...   ...  ...  ...  ...   
7560                           43, 45, 46, 47, 48, !=44   NaN  NaN  NaN  NaN   
7570                           43, 45, 46, 47, 48, !=44   NaN  NaN  NaN  NaN   
7793             40, !=65, !=66, !=67, !=68, !=69, !=70   NaN  NaN  NaN  NaN   
7937                                           40, !=49   NaN  NaN  NaN  NaN   
8041                                       52, 55, !=53   NaN  NaN  NaN  NaN   

        8                                               nums  \
95    NaN  [Primary Central Nervous System Lymphoma, Cent...   
107   NaN  [Primary Central Nervous System Lymphoma, Cent...   
118   NaN  [Primary Central Nervous System Lymphoma, Cent...   
125   NaN  [Primary Central Nervous System Lymphoma, Cent...   
185   NaN  [Primary Central Nervous System Lymphoma, Cent...   
...   ...                                                ...   
7560  NaN  [immunoblastic diffuse large b-cell lymphoma, ...   
7570  NaN  [immunoblastic diffuse large b-cell lymphoma, ...   
7793  NaN  [Non-Hodgkin Lymphoma, HIV related Lymphoma, a...   
7937  NaN  [Non-Hodgkin Lymphoma, Follicular Lymphoma (in...   
8041  NaN  [MALT-Lymphoma, gastric MALT, Marginal Zone Ly...   

                                              not_equal  
95    [nan, nan, spinal ependmyoma, nan, nan, nan, P...  
107   [nan, nan, spinal ependmyoma, nan, nan, nan, n...  
118   [nan, nan, spinal ependmyoma, nan, nan, nan, n...  
125   [nan, nan, spinal ependmyoma, nan, nan, nan, n...  
185   [nan, nan, spinal ependmyoma, nan, nan, nan, n...  
...                                                 ...  
7560  [nan, nan, nan, nan, Germinal Center Diffuse L...  
7570  [nan, nan, nan, nan, Germinal Center Diffuse L...  
7793  [nan, nan, nan, HIV related Lymphoma, nan, nan...  
7937  [nan, nan, nan, Follicular Lymphoma (includes ...  
8041                 [nan, nan, nan, nan, gastric MALT]  

[230 rows x 11 columns]

In [29]:
# Saving p_num_logic df into pickle file (To be used in making of master4 file) 
p_num_with_not_equal.to_pickle('p_num_with_not_equal.pkl')

In [24]:
# Create a copy of 'master2' to preserve the original DataFrame. The new DataFrame is named 'master3'.
master3 = master2.copy()

# Iterate through the values of columns '0' and 'not_equal' in the 'p_num_with_not_equal' DataFrame.
for rw, l in p_num_with_not_equal[[0, "not_equal"]].values:
    # Convert the 'not_equal' list to a pandas Series, drop NaN values, and convert it to a NumPy array.
    l = pd.Series(l).dropna().values
    
    # Create a temporary DataFrame 'df_tem' containing only the columns 'L1' to 'L5' from 'master2'.
    df_tem = master2.loc[:, "L1":"L5"]
    
    # Filter rows in 'df_tem' where any of the columns ('L1' to 'L5') contain values from the list 'l'.
    df_tem = df_tem[
        df_tem['L1'].isin(l) | 
        df_tem['L2'].isin(l) | 
        df_tem['L3'].isin(l) | 
        df_tem['L4'].isin(l) | 
        df_tem['L5'].isin(l)
    ]
    
    # Iterate through the index values of the filtered rows in 'df_tem'.
    for idx in df_tem.index.values:
        # Retrieve the 'Alternate_words' dictionary for the current index from 'master3'.
        d = master3.loc[idx, "Alternate_words"]
        
        # Initialize a new dictionary to store updated alternate words.
        d1 = {}
        
        # Check if the 'Alternate_words' value is a dictionary.
        if isinstance(d, dict):
            # Iterate through each key-value pair in the dictionary.
            for k in d:
                # Remove the word 'rw' from the alternate words list (case-insensitive) for the current key.
                dd1 = [str(i).strip() for i in d[k] if str(i).strip().lower() != str(rw).strip().lower()]
                
                # Remove empty strings from the cleaned list.
                dd1 = [i for i in dd1 if i != '']
                
                # If the cleaned list is not empty, add it to the updated dictionary.
                if len(dd1) > 0:
                    d1[k] = dd1
        
        # Update the 'Alternate_words' column in 'master3' for the current index with the cleaned dictionary.
        master3.at[idx, "Alternate_words"] = d1


In [25]:
master2=master3

In [26]:
# 
master2[master['index']==614]['Alternate_words'].values[0]

{'L1': ['Bone cancer (includes joint cancer and connective tissue cancer)',
  'cancer of any type',
  'solid-organ cancer',
  'malignant or nonmalignant disease',
  "pregnancy related' 'cancer",
  'extremities neoplasms',
  'malignant solid neoplasm',
  'any solid malignancy',
  'bone and joint cancer',
  'all  solid tumors',
  'non hematologic cancer',
  'childhood malignancy',
  'haemato-oncologic',
  'rare solid tumors',
  'any nonmyeloid tumor',
  'hematology and oncology',
  "'other solid tumors",
  'solid and hematological malignancies',
  'malignant disorders',
  'nonematologic malignancies',
  'other solid cancers',
  'all cancer',
  'tumors of any histology',
  'nonematological solid tumors',
  'neoplasms solid tumor',
  'solid tumor cancer',
  'solid tumours',
  'cancer nos',
  'pediatric, cancer',
  'cancer of bone and connective tissue',
  'solid malignancy',
  'malign neoplasm',
  'rare tumours',
  'malignant tumour',
  'malignant tumours of bone',
  'neoplasms by histolog

In [33]:
# Saving master df as master3.pkl
master2.to_pickle("master3.pkl")

In [34]:
# Examining the intermediary file
master2.to_excel("master2.xlsx",index=False)

## File Ends

### Above master3.pkl and p_num_with_not_equal.pkl files would be used in second sequel file of tumor_type module (serial # 2) . 
### **********************************